### PR TITLE
[Merged by Bors] - allow specifying pvc parameters

### LINKED ISSUE #67: Make resources configurable (CPU, memory, PVCs)
Especially StorageClass and size

This is done when
- [ ] resources are configurable according to the standards set by operator-rs
- [ ] documentation has been added
- [ ] integration test has been added

### PR DESCRIPTION
**Pull request by razvan**
## Description

Fixes #67 

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

**Comment by nightkr** [comment]
@soenkeliebau has been working on integrating something similar into his resource management scheme (over at https://github.com/stackabletech/operator-rs/compare/feat/resource_config2, don't think he made a proper PR from what I can tell?). Might be worth talking to him about aligning this.

**Comment by razvan** [comment]
> @soenkeliebau has been working on integrating something similar into his resource management scheme (over at https://github.com/stackabletech/operator-rs/compare/feat/resource_config2, don't think he made a proper PR from what I can tell?). Might be worth talking to him about aligning this.

Didn't know that. Thanks for the pointer.

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,7 +13,8 @@ All notable changes to this project will be documented in this file.
 - Config builder for `hdfs-site.xml` and `core-site.xml` ([#150]).
 - Discovery configmap that exposes the namenode services for clients to connect ([#150]).
 - Documented service discovery for namenodes ([#150]).
-- Publish warning events when role replicas don't meet certain minimum requirements ([#162])
+- Publish warning events when role replicas don't meet certain minimum requirements ([#162]).
+- PVCs for data storage, cpu and memory limits are now onfigurable ([#164]).
```
Comment (maltesander) [inline CHANGELOG.md]:
```suggestion
- PVCs for data storage, cpu and memory limits are now configurable ([#164]).
```

File: Cargo.toml [hunk 2]
```diff
@@ -3,3 +3,6 @@
 members = [
     "rust/crd", "rust/operator", "rust/operator-binary"
 ]
+
+#[patch."https://github.com/stackabletech/operator-rs.git"]
+#stackable-operator = { path = "../operator-rs" }
```
Comment (maltesander) [inline Cargo.toml]: remove?

File: docs/modules/ROOT/pages/usage.adoc [hunk 3]
```diff
@@ -163,3 +163,60 @@ nameNodes:
 ----
 
 IMPORTANT: Some environment variables will be overriden by the operator and cannot be set manually by the user. These are `HADOOP_HOME`, `HADOOP_CONF_DIR`, `POD_NAME` and `ZOOKEEPER`.
+
+=== Storage for data volumes
+
+You can mount volumes where data is stored by specifiying https://kubernetes.io/docs/concepts/storage/persistent-volumes[PersistentVolumeClaims] for each individual role group like so:
```
Comment (maltesander) [inline docs/modules/ROOT/pages/usage.adoc]:
```suggestion
You can mount volumes where data is stored by specifiying https://kubernetes.io/docs/concepts/storage/persistent-volumes[PersistentVolumeClaims] for each individual role group:
```

File: rust/crd/Cargo.toml [hunk 4]
```diff
@@ -8,7 +8,7 @@ repository = "https://github.com/stackabletech/hdfs-operator"
 version = "0.4.0-nightly"
 
 [dependencies]
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.15.0" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", branch = "feat/composable-merge" }
```
Comment (maltesander) [inline rust/crd/Cargo.toml]: Just a reminder to switch back on the tag (i guess 0.19.0?).

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -41,6 +46,21 @@ pub struct HdfsClusterSpec {
...
+            zookeeper_config_map_name: "zookeeper-config-map-for-hdfs".to_owned(),
+            data_nodes: None,
+            name_nodes: None,
+            journal_nodes: None,
+            dfs_replication: None,
+            log4j: None,
+        }
+    }
+}
+
```
Comment (maltesander) [inline rust/crd/src/lib.rs]: I dont get the benefit of that default? The only difference to deriving is pointing to a config map that does not exist? Whats the difference to the zookeeper_config_map_name defaulting to "" ?

File: rust/crd/src/lib.rs [hunk 6]
```diff
@@ -230,6 +250,130 @@ impl HdfsCluster {
...
+        .into_iter()
+        .flatten()
+        .reduce(|old, new| chainable_merge(new, old))
+        .unwrap(); // always succeeds because default_resources is Some()
+
+        let result_1 = r
+            .storage
+            .data
+            .build_pvc("data", Some(vec!["ReadWriteOnce"]));
+        let result_2 = r.clone().into();
```
Comment (maltesander) [inline rust/crd/src/lib.rs:268]: Could we get more readable variable names here?

File: rust/crd/src/lib.rs [hunk 7]
```diff
@@ -230,6 +250,130 @@ impl HdfsCluster {
...
+
+        (vec![result_1], result_2)
```
Comment (maltesander) [inline rust/crd/src/lib.rs:270]: I would not make this a vec just to fit the `volume_claim_templates`. Decreases testability etc.

File: rust/operator/src/hdfs_controller.rs [hunk 8]
```diff
@@ -302,23 +300,25 @@ fn rolegroup_statefulset(
     let init_containers;
     let containers;
 
+    let (pvc, rr) = hdfs.resources(role, rolegroup_ref);
```
Comment (maltesander) [inline rust/operator/src/hdfs_controller.rs]: Can we change the variable `rr` to `resources` or `resource_requirements`?

File: rust/crd/src/lib.rs [hunk 9]
```diff
@@ -511,3 +669,327 @@ impl Configuration for JournalNodeConfig {
...
+        replicas: 1
+  journalNodes:
+    roleGroups:
+      default:
+        selector:
+          matchLabels:
+            kubernetes.io/os: linux
+        replicas: 1
+
+        ";
```
Comment (maltesander) [inline rust/crd/src/lib.rs]: Is this for a linter or copy paste?
```suggestion
        replicas: 1";
```

File: rust/crd/src/lib.rs [hunk 10]
```diff
@@ -230,6 +250,130 @@ impl HdfsCluster {
...
+                        .map(|rg| &rg.config.config)
+                })
+                .and_then(|node_config| node_config.resources.clone()),
+        }
+    }
+
+    fn role_resources(
+        &self,
+        role: &HdfsRole,
+        _rolegroup_ref: &RoleGroupRef<HdfsCluster>,
```
Comment (maltesander) [inline rust/crd/src/lib.rs]: remove `_rolegroup_ref`?

File: rust/crd/src/lib.rs [hunk 11]
```diff
@@ -230,6 +250,130 @@ impl HdfsCluster {
...
+                .as_ref()
+                .map(|role| &role.config.config)
+                .and_then(|node_config| node_config.resources.clone()),
+        }
+    }
+
+    fn default_resources(
+        &self,
+        _role: &HdfsRole,
+        _rolegroup_ref: &RoleGroupRef<HdfsCluster>,
```
Comment (maltesander) [inline rust/crd/src/lib.rs]: remove?

**Review by maltesander — COMMENTED** [review]
Just some comments. Ill test tomorrow. Can we have a readme for the kuttl tests? Its not explained how to properly run that with the test definition yaml?

File: rust/operator/src/hdfs_controller.rs (hunk 8 above)
Comment (razvan) [reply rust/operator/src/hdfs_controller.rs]: Done.

File: rust/crd/src/lib.rs (hunk 10 above)
Comment (razvan) [reply rust/crd/src/lib.rs]: Done.

File: rust/crd/src/lib.rs (hunk 11 above)
Comment (razvan) [reply rust/crd/src/lib.rs]: Done.

File: rust/crd/src/lib.rs (hunk 7 above)
Comment (razvan) [reply rust/crd/src/lib.rs:270]: I don't see how this decreases testability.

File: rust/crd/src/lib.rs (hunk 5 above)
Comment (razvan) [reply rust/crd/src/lib.rs]: It was a leftover.

File: rust/crd/Cargo.toml (hunk 4 above)
Comment (razvan) [reply rust/crd/Cargo.toml]: After that branch is merged and a new release is done.

File: rust/crd/src/lib.rs (hunk 7 above)
Comment (maltesander) [reply rust/crd/src/lib.rs:270]: Ok let me rephrase then. Will there be more than one element in that vector in the future? If not, i dont see the point in returning a vec (simply to save on the `vec![xxx]`) when building the template?

File: Cargo.toml (hunk 2 above)
Comment (razvan) [reply Cargo.toml]: done.

File: rust/crd/src/lib.rs (hunk 7 above)
Comment (razvan) [reply rust/crd/src/lib.rs:270]: HDFS supports multiple volumes for it's data folders, so we might support that in the future. But to me, this is about encapsulation: I have one clear interface (this function) that returns PVCs and don't have to fiddle with them later in the code to pass them on to K8S.

Comment (maltesander) [reply rust/crd/src/lib.rs:270]: Ok did not know that. Then i resolve this.

**Review by maltesander — COMMENTED** [review]
Tested and works. @soenkeliebau should approve once he had a chance to look over this.
Locally the tests failed the first time (timeout 5min) and worked after the images were downloaded. Maybe we can increase the timeout for the install step?

**Review by soenkeliebau — APPROVED** [review]
I have looked it over and tested it, looks good to me.
One thing I noticed is that  the tests specify resources but don't check if they get set in the pods / pod templates correctly. That migth be an improvement down the line that we could add.
But I am also happy for this to go in as is.

**Comment by razvan** [comment]
bors merge
